lib/src/arithmetic/booth.dart

```diff
@@ -0,0 +1,744 @@
+
+    final lastRow = rows - 1;
+    final firstAddend = partialProducts[0];
+    final lastAddend = partialProducts[lastRow];
+    var alignRow0Sign = selector.width -
+        shift * lastRow -
+        ((shift > 1)
+            ? 1
+            : signed
+                ? 1
+                : 0);
+
+    if (alignRow0Sign < 0) {
+      alignRow0Sign = 0;
+    }
+
+    final signs = [for (var r = 0; r < rows; r++) encoder.getEncoding(r).sign];
+
+    final propagate =
+        List.generate(rows, (i) => List.filled(0, Logic(), growable: true));
+    for (var row = 0; row < rows; row++) {
+      propagate[row].add(signs[row]);
+      for (var col = 0; col < 2 * (shift - 1); col++) {
+        propagate[row].add(partialProducts[row][col]);
+      }
+      for (var col = 1; col < propagate[row].length; col++) {
+        propagate[row][col] = propagate[row][col] & propagate[row][col - 1];
+      }
+    }
+    final m =
+        List.generate(rows, (i) => List.filled(0, Logic(), growable: true));
+    for (var row = 0; row < rows; row++) {
+      for (var c = 0; c < shift - 1; c++) {
+        m[row].add(partialProducts[row][c] ^ propagate[row][c]);
+      }
+      m[row].addAll(List.filled(shift - 1, Logic()));
+    }
+
+    for (var i = shift - 1; i < m[lastRow].length; i++) {
+      m[lastRow][i] = lastAddend[i] ^
+          (i < alignRow0Sign ? propagate[lastRow][i] : Const(0));
+    }
+
+    final remainders = List.filled(rows, Logic());
+    for (var row = 0; row < lastRow; row++) {
+      remainders[row] = propagate[row][shift - 1];
+    }
+    remainders[lastRow] <= propagate[lastRow][alignRow0Sign];
+
+    // Compute Sign extension for row==0
+    final firstSign = signed ? firstAddend.last : signs[0];
+    final q = [
+      firstSign ^ remainders[lastRow],
+      ~(firstSign & ~remainders[lastRow]),
+    ];
+    q.insertAll(1, List.filled(shift - 1, ~q[1]));
+
+    for (var row = 0; row < rows; row++) {
+      final addend = partialProducts[row];
+      if (row > 0) {
+        final mLimit = (row == lastRow) ? 2 * (shift - 1) : shift - 1;
+        for (var i = 0; i < mLimit; i++) {
+          addend[i] = m[row][i];
+        }
+        // Stop bits
+        if (signed) {
+          addend.last = ~addend.last;
+        } else {
+          addend.add(~signs[row]);
+        }
+        addend
+          ..insert(0, remainders[row - 1])
+          ..addAll(List.filled(shift - 1, Const(1)));
+        rowShift[row] -= 1;
+      } else {
+        for (var i = 0; i < shift - 1; i++) {
+          firstAddend[i] = m[0][i];
+        }
+        if (signed) {
+          firstAddend.last = q[0];
+        } else {
+          firstAddend.add(q[0]);
+        }
+        firstAddend.addAll(q.getRange(1, q.length));
+      }
+    }
+    if (shift == 1) {
+      lastAddend.add(Const(1));
+    }
+  }
+
+  /// Sign extend the PP array using stop bits without adding a row
+  /// This routine works with different widths of multiplicand/multiplier
```

Comment:
I added a reference for the base implementation.  I think I have publishable innovation on top which is to make it work for

1. Widths that are not multiples of the shift of the radix (e.g. shift of radix 8 is 3).  This was the first problem I discovered in the publication that they had conveniently chosen w=12 because it was a multiple of the shift and didn't work at all widths.
2. rectangular multipliers where the multiplier and multiplicand are different widths (I see that as a potential in ML dot-products -- perhaps a convenience, but perhaps that flexibility will become quite useful.
3. signed vs unsigned: the paper did not address, and it turns out to be interestingly tricky. I found that a Stanford thesis appendix which made it clear that you have to extend the multiplier when signed, not obvious at all!